SoftwareGuide/Latex/Architecture/SystemOverview.tex

```diff
@@ -378,27 +378,96 @@ \subsection{Multi-Threading}
+to update filter's progress after each region is finished processing.
+
+If a filter has some serial part in the middle, in addition to initialization
+done in \code{BeforeThreadedGenerateData()} and finalization done in
+\code{AfterThreadedGenerateData()}, it can parallelize more than one method
+in its own version of \code{GenerateData()}, such as done by
+\doxygen{CannyEdgeDetectionImageFilter}:
+
+\small
+\begin{minted}[baselinestretch=1,fontsize=\footnotesize,linenos=false,bgcolor=ltgray]{cpp}
+::GenerateData()
+{
+  this->UpdateProgress(0.0f);
+  Superclass::AllocateOutputs();
+  //small serial section
+  this->UpdateProgress(0.01f);
+
+  // Calculate 2nd order directional derivative
+  this->GetMultiThreader()->template ParallelizeImageRegion<TOutputImage::ImageDimension>(
+      this->GetOutput()->GetRequestedRegion(),
+      [this](const OutputImageRegionType & outputRegionForThread)
+      { this->ThreadedCompute2ndDerivative(outputRegionForThread); }, nullptr);
+  this->UpdateProgress(0.45f);
+
+  // Calculate the gradient of the second derivative
+  this->GetMultiThreader()->template ParallelizeImageRegion<TOutputImage::ImageDimension>(
+      this->GetOutput()->GetRequestedRegion(),
+      [this](const OutputImageRegionType & outputRegionForThread)
+      { this->ThreadedCompute2ndDerivativePos(outputRegionForThread); }, nullptr);
+  this->UpdateProgress(0.9f);
+
+  //more processing
+  this->UpdateProgress(1.0f);
+}
+\end{minted}
+\normalsize
+
+When invoking \code{ParallelizeImageRegion} multiple times from
+\code{GenerateData()}, \code{nullptr} should be passed instead of \code{this},
+otherwise progress will go from 0\% to 100\% more than once. And this
+will at least confuse any other class watching the filter's progress events,
+even if it does not cause a crash. So the filter's author should estimate
+how long each part of \code{GenerateData()} takes, and update the progress
+``manually'' as in the example above.
+
+With ITK version 5.0, the Multi-Threading mechanism has been refactored.
+What was previously \code{itk::MultiThreader}, is now a hierarchy of classes.
+\doxygen{PlatformMultiThreader} is a slightly cleaned-up version of the old
+class - \code{MultipleMethodExecute} and \code{SpawnThread} methods have been
+deprecated. But much of its content has been moved to
+\doxygen{MultiThreaderBase}. And classes should use the multi-threaders via
+\code{MultiThreaderBase} interface, to allow the end user the flexibility to
+select the multi-threader at run time. This also allows the filter to
+benefit from future improvements in threading such as addition of a new
+multi-threader implementation.
+
+The backwards compatible \code{ThreadedGenerateData(Region, ThreadId)} method
+signature has been kept, for use in filters that must know their thread number.
+To use this signature, a filter must invoke
+\code{this->DynamicMultiThreadingOff();} prior to \code{Update();}. The best
```

Comment:
`(...) prior to \code{this->Update();}`

for consistency and conform to the style guidelines.

Reply:
Filters don't usually invoke Update() on themselves, rather the user of the filter does that. Hence no `this->` prefix. I will reword, though.